### DISCUSSION TITLE
It feels like the return value of ntop is not a string

### DISCUSSION THREAD

**Discussion by dongbiao830415** [bpftrace/bpftrace#3582] [Q&A]
#!/usr/bin/env bpftrace

#include <linux/netfilter.h>

uprobe:/opt/WiseGrid/smartlb/smartlb:vmac_check_enable {
    $family = (uint16)arg0;
    $addr = (union nf_inet_addr *)arg1;

    if ($family == AF_INET) {
        $ip = ntop(AF_INET, $addr->ip);
    } else {
        $ip = ntop(AF_INET6, $addr->ip6);
    }

    @tip[tid] = $ip; //expect strings to be stored in @tip[tid]
    if (@tip[tid] != "") { //ERROR: Type mismatch for '!=': comparing 'inet' with 'string[1]';  Why?
        printf("some message\n");
    }
    printf("%d\t%s\t%s\n", tid, $ip, @tip[tid]);//what prints out here is a string
}

bpftrace version :bpftrace v0.18.1-2-g7059

**Comment by jordalgo** [comment]
Unfortunately `ntop` returns an internal bpftrace type `inet`, which is why you're seeing the type mismatch and at the moment you can't cast either one to the other. 

We have a fix for the map lookup on master (unreleased) as we introduced the `has_key` helper: https://github.com/bpftrace/bpftrace/blob/master/man/adoc/bpftrace.adoc#has_key

That being said, from your code it looks like the map lookup will never be null or empty as you're setting it right before the check - that is of course unless either of the ntop calls return an empty-ish `inet`.

**Comment by ajor** [comment]
Might it be useful to have a way to create an `inet` type from a string literal, so that it could be used in comparisons?

e.g. `if (@tip[tid] != inet("127.0.0.1"))`

**Reply by jordalgo** [reply]
I like it or maybe we could do a cast?
